docs/source/traits_user_manual/internals.rst

```diff
@@ -0,0 +1,190 @@
+    {}
+    >>> eggs._class_traits()
+    {
+        'name': <traits.ctrait.CTrait object at 0x1020cd400>,
+        'quantity': <traits.ctrait.CTrait object at 0x1020cd360>,
+        'trait_added': <traits.ctrait.CTrait object at 0x1020cd4a0>,
+        'trait_modified': <traits.ctrait.CTrait object at 0x1020079a0>,
+    }
+    >>> eggs.__dict__
+    {'name': 'eggs', 'quantity': 12.0}
+
+Note that the actual values for the ingredient are stored in the ``__dict__``
+as usual, not in the ``CTrait`` objects.
+
+If we create a second ingredient, it shares class traits (but not
+instance traits) with the first one::
+
+    >>> flour = Ingredient(name="flour", quantity=3.5)
+    >>> flour._class_traits() is eggs._class_traits()
+    True
+    >>> flour._instance_traits() is eggs._instance_traits()
+    False
+
+
+Key class: |CTrait|
+~~~~~~~~~~~~~~~~~~~
+
+A |CTrait| object has two main purposes:
+
+- It encapsulates the rules for getting and setting a traited attribute on
+  a |HasTraits| object.
+- It provides an attachment point for trait notifiers.
+
+For attribute access, the C structure underlying each |CTrait| instance (see
+the ``trait_object`` struct in the C source) has three relevant fields:
+``getattr``, ``setattr`` and ``post_setattr``. These fields contain C function
+pointers, and are invoked during attribute get and set operations, as described
+below.
+
+
+Attribute retrieval
+~~~~~~~~~~~~~~~~~~~
+
+When evaluating ``obj.name`` for a ``HasTraits`` object ``obj``, the following
+sequence of steps occurs (see ``has_traits_getattro`` in the C source):
+
+- The ``name`` is looked up in ``obj.__dict__``. If found, the corresponding
+  value is returned immediately.
+- If ``name`` is not found in ``obj.__dict__``, we look first for an instance
+  trait named ``name``, and then for a class trait named ``name``. Thus an
+  instance trait with a given name will shadow a class trait with the same
+  name.
+- If a matching trait is found, its ``getattr`` field is invoked to retrieve
+  the trait's value for the given object.
+- If no matching trait is found, we try to access the attribute value
+  using Python's usual attribute rules (via the ``PyObject_GenericGetAttr``
+  C-API call).
+- Finally, if the ``PyObject_GenericGetAttr`` call fails, we invoke the
+  **prefix trait** machinery to get a new ``CTrait`` object, and use that
+  new trait to get a value.
+
+Note that the above sequence of steps applies to method access as well as
+attribute access. Note also that there's no mechanism to automatically
+search for ``CTrait`` objects in superclasses of the immediate ``HasTraits``
+subclass.
+
+
+Attribute set operations
+~~~~~~~~~~~~~~~~~~~~~~~~
+
+The rules for setting an attribute (evaluating ``obj.name = value`` for a
+``HasTraits`` object ``obj``) are analogous to those for attribute retrieval.
+The starting point is ``has_traits_setattro`` in the source.
+
+- First we look for the name ``name`` in ``obj``s instance traits dictionary,
```

Comment:
Nit: `obj`'s (missing an apostrophe) ?
The formatting gets a bit upset here (as shown on GitHub and with rst2html). Looks like we need an extra space between `obj` and "'s".
Same applies to the bullet point below.

Reply:
I reworded to avoid the awkward possessive. See what you think.